pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -478,6 +470,15 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
 		return nil, err
 	}
 
+	attachablePlugin, err := rc.volumePluginMgr.FindAttachablePluginBySpec(volumeSpec)
```

Comment:
@dobsonj Could you explain how this change fix the issue? I guess that this makes`GetUniqueVolumeNameFromSpecWithPod` called instead of `GetUniqueVolumeNameFromSpec`, but I can't see why (and whether it is the right way).

Reply:
@mkimuram That's right, `GetUniqueVolumeNameFromSpec` is valid only when a PV is attached. I brought this up with Jan and Hemant, and they pointed me to this code in DSW where we do a similar check but we find the plugin by spec instead of by name:
https://github.com/kubernetes/kubernetes/blob/5e63432f79357c376d920ae35409e98502bec6ec/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L240-L243
If we use [FindAttachablePluginByName](https://github.com/kubernetes/kubernetes/blob/5e63432f79357c376d920ae35409e98502bec6ec/pkg/volume/plugins.go#L904-L917) here, it will find the plugin even for ephemeral volume times, which means `attachablePlugin != nil` and then we call `GetUniqueVolumeNameFromSpec`.
But if we use [FindAttachablePluginBySpec](https://github.com/kubernetes/kubernetes/blob/5e63432f79357c376d920ae35409e98502bec6ec/pkg/volume/plugins.go#L885-L902), it can check if the volume is actually attachable with [CanAttach](https://github.com/kubernetes/kubernetes/blob/5e63432f79357c376d920ae35409e98502bec6ec/pkg/volume/csi/csi_plugin.go#L612), which returns false for ephemeral volume times (i.e. CSI inline volumes).
After this change, `attachablePlugin` and `deviceMountablePlugin` will be nil for ephemeral volume types (i.e. no PV to attach) when we reach this check:
https://github.com/kubernetes/kubernetes/blob/6fcbb2c64eb07564bfb801f33c84cac01f59900b/pkg/kubelet/volumemanager/reconciler/reconciler.go#L483-L490

Reply:
Thank you for your explanation. I understand that the previous code just checks if the plugin is attachable and the current code checks if the volume can actually be attached. As "`GetUniqueVolumeNameFromSpec` is valid only when a PV is attached.", it seems more accurate to check like the current code. If it doesn't affect other plugins and features (I hope that the test results eventually imply that it's true), this approach looks good.

Reply:
Could you summarize it and add it to comments why using spec to get plugin instead name?